azure-servicebus/src/main/java/com/microsoft/azure/servicebus/primitives/RequestResponseLink.java

```diff
@@ -147,6 +167,28 @@ public String getLinkPath()
 	    return this.linkPath;
 	}
 	
+	private CompletableFuture<Void> sendTokenAndSetRenewTimer(boolean retryOnFailure)
+    {
+        if(this.getIsClosingOrClosed() || this.sasTokenAudienceURI == null)
+        {
+            return CompletableFuture.completedFuture(null);
+        }
+        else
+        {
+            CompletableFuture<ScheduledFuture<?>> sendTokenFuture = this.underlyingFactory.sendSecurityTokenAndSetRenewTimer(this.sasTokenAudienceURI, retryOnFailure, () -> this.sendTokenAndSetRenewTimer(true));
+            return sendTokenFuture.thenAccept((f) -> {this.sasTokenRenewTimerFuture = f; TRACE_LOGGER.debug("Set SAS Token renew timer");});
+        }
+    }
+	
+	private void cancelSASTokenRenewTimer()
+    {
+        if(this.sasTokenRenewTimerFuture != null && !this.sasTokenRenewTimerFuture.isDone())
+        {
+            TRACE_LOGGER.debug("Cancelling SAS Token renew timer");
+            this.sasTokenRenewTimerFuture.cancel(true);
```

Comment:
>this.sasTokenRenewTimerFuture.cancel(true); [](start = 12, length = 43)

There could potentially be a race condition between this and 
`sendTokenFuture.thenAccept((f) -> {this.sasTokenRenewTimerFuture = f;`

Should we take care of that?

Reply:
Cancelling timer always is always a race. The timer may fire while cancellation is in progress. Or cancellation may happen while timer action is already triggered. But the result of the race is largely harmless. Even if the timer fires after cancellation, it will check if the object is closed or not before proceeding to do anything.

---
In reply to: [170118663](https://github.com/Azure/azure-service-bus-java/pull/188#discussion_r170118663) [](ancestors = 170118663)